### PR TITLE
#5811 Refactor for Magic Commands.

### PR DESCRIPTION
**Pull request by michalgce**

### CODE REVIEW

**Comment by michalgce** [comment]
@scottdraves add_magic

File: kernel/base/src/main/java/com/twosigma/beakerx/kernel/commands/CommandExecutorImpl.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    this.messageCreator = new MessageCreator(kernel);
+    this.availableCommands = buildAvailableCommands();
+    availableCommands.add(new LSMagicCommand(availableCommands, messageCreator));
+    this.commandProcessor = new CommandProcessorImpl(availableCommands);
+    this.commandPublisher = new CommandPublisherImpl(kernel);
+  }
+
+  private List<MagicCommand> buildAvailableCommands() {
+    ArrayList<MagicCommand> magicCommands = Lists.newArrayList(new BashMagicCommand(messageCreator),
+                                                               new HTMLMagicCommand(messageCreator),
```
Comment (scottdraves) [inline kernel/base/src/main/java/com/twosigma/beakerx/kernel/commands/CommandExecutorImpl.java:64]: too much indentation

File: kernel/base/src/main/java/com/twosigma/beakerx/kernel/commands/CommandProcessorImpl.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+      List<MagicCommand> availableCommands) {
+    this.availableCommands = availableCommands;
+  }
+
+  @Override
+  public List<CommandItem> process(Message message, Integer executionCount) {
+    Code code = takeCodeFrom(message);
+
+    return buildCommandsMap(code).stream()
+                                 .map(commandCode -> commandCode.getCommand().build().process(commandCode.getCode(), message, executionCount))
```
Comment (scottdraves) [inline kernel/base/src/main/java/com/twosigma/beakerx/kernel/commands/CommandProcessorImpl.java:41]: line too long

File: kernel/base/src/main/java/com/twosigma/beakerx/kernel/commands/CommandProcessorImpl.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+  private List<CommandCode> buildCommandsMap(Code code) {
+    List<CommandCode> commandsToExecute = Lists.newArrayList();
+
+    for (String lineOfCode: code.asString().split("\\r\\n|\\n|\\r")) {
+      lineOfCode = lineOfCode.replaceAll("\\s+", " ");
+      if (isThatMagic(lineOfCode)) {
+        MagicCommand magicCommandAndCode = getMagicCommandAndCode(lineOfCode);
+        if (isThatMagicCell(lineOfCode)) {
+          commandsToExecute.add(new CommandCode(magicCommandAndCode, code.asString().replace(magicCommandAndCode.getName(), "").trim()));
```
Comment (scottdraves) [inline kernel/base/src/main/java/com/twosigma/beakerx/kernel/commands/CommandProcessorImpl.java:53]: line too long.  please keep all code visible in github without h scrolling.

File: kernel/base/src/main/java/com/twosigma/beakerx/kernel/commands/type/Command.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package com.twosigma.beakerx.kernel.commands.type;
+
+import com.twosigma.beakerx.kernel.commands.MagicCommandFunctionality;
+
+public interface Command {
```
Comment (scottdraves) [inline kernel/base/src/main/java/com/twosigma/beakerx/kernel/commands/type/Command.java:20]: this doesn't look right, each command should have its own definition.

**Comment by scottdraves** [comment]
1. ExecuteRequestHandler
  a) here https://github.com/twosigma/beakerx/blob/michal/5811/kernel/base/src/main/java/com/twosigma/beakerx/kernel/handler/ExecuteRequestHandler.java#L66  we lost synchronization, an action "syncObject.release();" should be called at the end of execution, on master we call this method in callback, I think It will cause a problem with "runAll" functionality

2. CommandExecutor  the name is misleading, commandExecutor tells me that we execute commands so we should pass here a command not message


2. CommandExecutorImpl
a) https://github.com/twosigma/beakerx/blob/michal/5811/kernel/base/src/main/java/com/twosigma/beakerx/kernel/commands/CommandExecutorImpl.java
 CommandExecutorImpl shouldn't keep available magic commands

b) https://github.com/twosigma/beakerx/blob/michal/5811/kernel/base/src/main/java/com/twosigma/beakerx/kernel/commands/CommandExecutorImpl.java#L78
  shoudl be defined somewhere in sql kernel

c) generally this class could be removed and we could move the code from execute method to ExecuteRequestHandler

3. CommandPublisherImpl
  a) the name is misleading and it doesn't publish only messages but it executes code as well
https://github.com/twosigma/beakerx/blob/michal/5811/kernel/base/src/main/java/com/twosigma/beakerx/kernel/commands/CommandPublisherImpl.java#L41
additionally it creates callback (seo) -> kernel.sendIdleMessage(seo.getJupyterMessage()) which was ealier in ExecuteRequestHandler 
https://github.com/twosigma/beakerx/blob/master/kernel/base/src/main/java/com/twosigma/beakerx/kernel/handler/ExecuteRequestHandler.java#L115

File: kernel/base/src/main/java/com/twosigma/beakerx/kernel/commands/type/ClassPathMagicCommand.java [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+    return new ErrorData(false, "");
+  }
+
+  private Boolean doesPathContainsWildCards(String path) {
+    return path.contains("*");
+  }
+
+  private Boolean containsSingleWildcardSymbol(String path) {
+    return path.length() - path.replace("*", "").length() == 1;
```
Comment (scottdraves) [inline kernel/base/src/main/java/com/twosigma/beakerx/kernel/commands/type/ClassPathMagicCommand.java:131]: too complicated.

File: kernel/base/src/main/java/com/twosigma/beakerx/kernel/commands/type/CellTimeItMagicCommand.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+  public CellTimeItMagicCommand(KernelFunctionality kernel, MessageCreator messageCreator) {
+    super(TIMEIT_CELL, "", Sets.newHashSet(MagicCommandType.CELL), messageCreator, kernel);
+  }
+
+  @Override
+  public MagicCommandFunctionality build() {
+    return (code, message, executionCount) -> {
+      try {
+        String codeToExecute = code.replaceAll("(-.)(\\d*\\s)", "");
```
Comment (scottdraves) [inline kernel/base/src/main/java/com/twosigma/beakerx/kernel/commands/type/CellTimeItMagicCommand.java:34]: what does this do?  too complicated.

File: kernel/base/src/main/java/com/twosigma/beakerx/kernel/commands/type/LineTimeItMagicCommand.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+  public LineTimeItMagicCommand(KernelFunctionality kernel, MessageCreator messageCreator) {
+    super(TIMEIT_LINE, "", Sets.newHashSet(MagicCommandType.LINE), messageCreator, kernel);
+  }
+
+  @Override
+  public MagicCommandFunctionality build() {
+    return (code, message, executionCount) -> {
+      String codeToExecute = code.replace(TIMEIT_LINE, "");
+      codeToExecute = codeToExecute.replaceAll("(-.)(\\d*\\s)", "");
```
Comment (scottdraves) [inline kernel/base/src/main/java/com/twosigma/beakerx/kernel/commands/type/LineTimeItMagicCommand.java:41]: i saw this before, 2 places makes it even worse.

**Comment by scottdraves** [comment]
too many try blocks.  almost every one should be done with regular control flow.

**Pull request closed without merging**